Scheduler plugin test improvements

lukeseawalker opened the pull request:
### Description of changes
* Merge from develop
* Remove test env values
* Fix release for shared session fixture
  * Only the shared fixture owner must be able to release the fixture data
* Add new fixtures to handle bucket at session level
  * session_s3_bucket_factory create a bucket factory at session level
  * s3_buckets is a fixture shared across multiple pytest-xdist processes, that permit to have just 1 test bucket for each region
  * s3_bucket retrieve the bucket for a given region and can be used at class level in the tests
* Switch a test to use the new single bucket per region
* Fix register cli credentials method
  * Check initialize_cli_creds is empty before going through creds initialization
* Use new scheduler commands factory fixture
  * Modify tests to use the new scheduler commands factory fixture in place of calling directly the get_scheduler_commands method
* Enable local path for scheduler-definition
* Fix slurm command launch
  * slurm-user requires special sudoers configuration to be able to launch slurm commands as fleet-mgt-user.
    This configuration is not manageable through SudoerConfiguration of SystemUsers.
    The soudoers configuration is moved under a recipe template
* Change config for test slurm compatible with scheduler-plugin
* Add configurable bucket prefix for template plugin definition
* Fix xdist_session_fixture for fixture with yield
* Improve fixtures to share objects at sessions level
* Change the upload of scheduler plugin definition as a separated fixture
* Remove path from slurm commands
  * Commands are in the path because the profile with the right slurm install location is loaded during user login.
     For the sudo case, it's necessary to have a login shell 


### Tests
Tests performed with the following config

```
{%- import 'common.jinja2' as common with context -%}
---
test-suites:
  schedulers:
    test_slurm.py::test_slurm:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["my_plugin"]
scheduler-plugins:
  my_plugin:
    scheduler-definition: "../../scheduler_plugins/slurm/utils/upload_artifacts.sh"
    scheduler-commands: "tests.common.schedulers_common.SlurmCommands"
    requires-sudo: true
```

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

demartinofra committed:
[integ-tests] Implement possibility to run tests against scheduler plugins

demartinofra committed:
[integ-tests] Implement possibility to run tests against scheduler plugins

demartinofra committed:
[integ-tests] Add fixture that can be shared across xdist processes

Signed-off-by: Francesco De Martino <fdm@amazon.com>

lukeseawalker committed:
Remove test env values

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Remove test env values

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Fix release for shared session fixture

Only the shared fixture owner must be able to release the fixture data

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Add new fixtures to handle bucket at session level

* session_s3_bucket_factory create a bucket factory at session level
* s3_buckets is a fixture shared across multiple pytest-xdist processes, that permit to have just 1 test bucket for each region
* s3_bucket retrieve the bucket for a given region and can be used at class level in the tests

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Switch test to use the new single bucket per region

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Fix register cli credentials method

Check initialize_cli_creds is empty before going through creds initialization

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Use new scheduler commands factory fixture

Modify tests to use the new scheduler commands factory fixture in place of calling directly the get_scheduler_commands method

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Enable local path for scheduler-definition

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Fix slurm command launch

slurm-user requires special sudoers configuration to be able to launch slurm commands as fleet-mgt-user.
This configuration is not manageable through SudoerConfiguration of SystemUsers.
The soudoers configuration is moved under a recipe template

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Fix line too long

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Skip complexity check for inject_additional_config_settings function

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Change config for test slurm compatible with scheduler-plugin

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Add configurable bucket prefix for template plugin definition

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

demartinofra commented on tests/integration-tests/README.md:
```diff
@@ -861,3 +887,82 @@ def run_benchmarks(request, mpi_variants, test_datadir, instance, os, region, be
 Please remove `IdentityFile` option from `ssh/config` before running the testing framework
+
+
+## Testing a scheduler plugin
```
Do you want to mention the possibility to use a local plugin?

lukeseawalker replied:
it's done when describing the `scheduler-definition`. Were you thinking to something different?

demartinofra commented on tests/integration-tests/conftest.py:
```diff
@@ -814,7 +848,7 @@ def get_availability_zones(region, credential):
 
-@pytest.fixture(scope="session", autouse=True)
+@xdist_session_fixture(autouse=True)
 def initialize_cli_creds(cfn_stacks_factory, request):
```
todo: remove dep on `cfn_stacks_factory` session fixture

demartinofra commented on tests/integration-tests/conftest.py:
```diff
@@ -1022,6 +1065,78 @@ def _create_bucket():
 
 
+@xdist_session_fixture(autouse=True)
+def s3_buckets(session_s3_bucket_factory, request):
```
remove dep on `session_s3_bucket_factory` session fixture and move code in here

demartinofra commented on tests/integration-tests/conftest.py:
```diff
@@ -1022,6 +1065,78 @@ def _create_bucket():
+    s3_buckets_dict = {}
+    for region in regions:
+        s3_bucket = session_s3_bucket_factory(region)
+        scheduler_definitions_dict = _upload_scheduler_plugin_definitions(request, s3_bucket, region)
```
we could transform `upload_scheduler_plugin_definitions` into a xdist_session_fixture that depends on s3_buckets just to keep the two logic separate

lukeseawalker replied:
done

demartinofra commented on tests/integration-tests/conftest.py:
```diff
@@ -1022,6 +1065,78 @@ def _create_bucket():
+                logging.info(
+                    "Found scheduler definition (%s) for scheduler plugin (%s)", scheduler_definition, plugin_name
+                )
+                _is_url(scheduler_definition)
```
do we need this? can we do validation in config_validator?

lukeseawalker replied:
removed, validation will not be part of this PR

demartinofra commented on tests/integration-tests/tests/ad_integration/test_ad_integration.py:
```diff
@@ -601,7 +599,7 @@ def test_ad_integration(
     )
 
-    scheduler_commands = get_scheduler_commands(scheduler, remote_command_executor)
+    scheduler_commands = scheduler_commands_factory(remote_command_executor)
```
can you please check the Code scanning error below?

demartinofra commented on tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py:
```diff
@@ -632,7 +633,7 @@ def get_config_param_vals():
 
 
-def test_cloudwatch_logging(region, scheduler, instance, os, pcluster_config_reader, test_datadir, clusters_factory):
+def test_cloudwatch_logging(region, scheduler, os, pcluster_config_reader, test_datadir, clusters_factory):
```
can you please check the Code scanning error below?

lukeseawalker committed:
Fix xdist_session_fixture for fixture with yield

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Improve fixtures to share objects at sessions level

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Change the upload of scheduler plugin definition as a separated fixture

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Remove path from slurm commands

Commands are in the path because the profile with the right slurm install location is loaded during user login.
For the sudo case, it's necessary to have a login shell

Signed-off-by: Luca Carrogu <carrogu@amazon.com>
